### PR TITLE
Introduction of the `[MultipleOf]` validation attribute

### PR DESCRIPTION
**Pull request by Corniel**
One of the constraints that can be defined in Open API models is `mutlipleof`. To generate models that fully support that concept, a `[MultipleOf]` attribute comes in handy.

### CODE REVIEW

File: specs/Qowaiv.Validation.Specs/DataAnnotations/Attributes/Multiple_of_specs.cs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    [TestCase("45.500%", 1.00)]
+    [TestCase("45.120%", 0.10)]
+    [TestCase("45.178%", 0.01)]
+    public void percentages(Percentage value, double factor)
+       => new MultipleOfAttribute(factor).IsValid(value).Should().BeFalse();
+}
+
+public class With_message
+{
+    [TestCase("nl", "De waarde van veld Total field is geen veelvoud van 0,001.")]
```
Comment (LauraXiulan) [inline specs/Qowaiv.Validation.Specs/DataAnnotations/Attributes/Multiple_of_specs.cs:136]: 'field' can be removed

File: specs/Qowaiv.Validation.Specs/DataAnnotations/Attributes/Multiple_of_specs.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            .WithMessage(ValidationMessage.Error(message, "Total"));
+    }
+
+    internal class Model
+    {
+        [MultipleOf(0.001)]
+        public Amount Total { get; set; } = 0.12345.Amount();
+
+        [MultipleOf(0.001)]
+        public Amount Sub { get; set; } = 0.123.Amount();
```
Comment (LauraXiulan) [inline specs/Qowaiv.Validation.Specs/DataAnnotations/Attributes/Multiple_of_specs.cs:151]: Is Sub used anywhere?

File: src/Qowaiv.Validation.DataAnnotations/Qowaiv.Validation.DataAnnotations.csproj [hunk 3]
```diff
@@ -60,9 +64,7 @@ v0.0.4
   </ItemGroup>
 
   <ItemGroup>
-    <EmbeddedResource Update="QowaivValidationMessages.nl.resx">
-      <Generator>ResXFileCodeGenerator</Generator>
-    </EmbeddedResource>
+    <EmbeddedResource Update="QowaivValidationMessages.nl.resx" />
```
Comment (LauraXiulan) [inline src/Qowaiv.Validation.DataAnnotations/Qowaiv.Validation.DataAnnotations.csproj:67]: You don't need the generator for the dutch messages anymore?

File: src/Qowaiv.Validation.DataAnnotations/Attributes/MultipleOfAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    /// <inheritdoc />
+    [Pure]
+    public override bool IsValid(object? value) => value switch
+    {
+        float flt => IsValid(flt),
+        double dbl => IsValid(dbl),
+        decimal dec => IsValid(dec),
+        Amount amount => IsValid((decimal)amount),
+        Money money => IsValid((decimal)money.Amount),
+        Percentage perentage => IsValid(100 * (decimal)perentage),
```
Comment (LauraXiulan) [inline src/Qowaiv.Validation.DataAnnotations/Attributes/MultipleOfAttribute.cs:48]: percentage

File: specs/Qowaiv.Validation.Specs/DataAnnotations/Attributes/Multiple_of_specs.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    [TestCase("45.000%", 1.000)]
+    [TestCase("45.100%", 0.100)]
+    [TestCase("45.170%", 0.010)]
+    [TestCase("45.178%", 0.001)]
+    public void percentages(Percentage value, double factor)
+       => new MultipleOfAttribute(factor).IsValid(value).Should().BeTrue();
+
+    [TestCase(nameof(NonFloatingPoints))]
+    public void non_floating_points(object model)
+      => new MultipleOfAttribute(10).IsValid(model).Should().BeTrue();
```
[the remark this answers is not in the document]
Comment (Corniel) [reply specs/Qowaiv.Validation.Specs/DataAnnotations/Attributes/Multiple_of_specs.cs:58]: Fair point: a check in some MS implementations suggest that you should return false for not supported values.

File: src/Qowaiv.Validation.DataAnnotations/Qowaiv.Validation.DataAnnotations.csproj (hunk 3 above)
Comment (Corniel) [reply src/Qowaiv.Validation.DataAnnotations/Qowaiv.Validation.DataAnnotations.csproj:67]: Never needed.

File: src/Qowaiv.Validation.DataAnnotations/Attributes/MultipleOfAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        _ => true,
```
Comment (LauraXiulan) [inline src/Qowaiv.Validation.DataAnnotations/Attributes/MultipleOfAttribute.cs:49]: So int is always valid?

File: src/Qowaiv.Validation.DataAnnotations/README.md [hunk 7]
```diff
@@ -1 +1,128 @@
...
+
+``` C#
+var validator = new AnnotatedModelValidator<SomeModel>();
+Result<SomeModel> result = validator.Validate(model);
+```
+
+## Validation messages
+`Qowaiv.Validation.DataAnnotations.ValidationMessage` inherits Microsoft's
+`ValidationResult`, and implements `Qowaiv.Validation.Abstractions.IValidationMessage`.
+This allows the creates messages with different severities:
```
Comment (LauraXiulan) [inline src/Qowaiv.Validation.DataAnnotations/README.md:17]: This allows the **creation of** messages with different severities:

File: src/Qowaiv.Validation.DataAnnotations/README.md [hunk 8]
```diff
@@ -1 +1,128 @@
...
+    public EmailAddress Email { get; set; }
+
+    [Mandatory()]
+    public string SomeString { get; set; }
+}
+```
+
+### Any
+The `[Required]` attribute does not work (well) for collections. The `[Any]`
+attribute does. It is only valid as he collection contains at least one item.
```
Comment (LauraXiulan) [inline src/Qowaiv.Validation.DataAnnotations/README.md:48]: It is only valid **if the** collection contains at least one item.

File: src/Qowaiv.Validation.DataAnnotations/README.md [hunk 9]
```diff
@@ -1 +1,128 @@
...
+``` C#
+public class Model
+{
+    [Any()]
+    public List<int> Numbers { get; set; }
+}
+```
+
+### Allowed values
+The `]AllowedValues]` attribute allows to define a subset of allowed values. It
```
Comment (LauraXiulan) [inline src/Qowaiv.Validation.DataAnnotations/README.md:59]: Bracket the wrong way

File: src/Qowaiv.Validation.DataAnnotations/README.md [hunk 10]
```diff
@@ -1 +1,128 @@
...
+public class Model
+{
+    [AllowedValues("DE", "FR", "GB")]
+    public Country CountryOfBirth { get; set; }
+}
+```
+
+### Forbidden values
+The `[ForbiddenValues]` attribute allows to define a subset of forbidden values. It
+supports type converters to get the allowed values based on a string value.
```
Comment (LauraXiulan) [inline src/Qowaiv.Validation.DataAnnotations/README.md:72]: forbidden values

Comment (LauraXiulan) [reply src/Qowaiv.Validation.DataAnnotations/README.md:72]: Not resolved yet

File: src/Qowaiv.Validation.DataAnnotations/README.md [hunk 11]
```diff
@@ -1 +1,128 @@
...
```
Comment (LauraXiulan) [inline src/Qowaiv.Validation.DataAnnotations/README.md:72]: This line still says allowed values instead of forbidden values

**Review by LauraXiulan — APPROVED** [review]

File: src/Qowaiv.Validation.DataAnnotations/README.md (hunk 11 above)
Comment (Corniel) [reply src/Qowaiv.Validation.DataAnnotations/README.md]: Ah, that was what you ment. :)

**Pull request merged**